Box.V2/JWTAuth/BoxJWTAuth.cs

```diff
@@ -145,27 +149,61 @@ public string UserToken(string userId)
 
         private string GetToken(string subType, string subId)
         {
+            int retryCounter = 0;
+            ExponentialBackoff expBackoff = new ExponentialBackoff();
+
             var assertion = ConstructJWTAssertion(subId, subType);
             OAuthSession result;
-            try
-            {
-                result = JWTAuthPost(assertion);
-                return result.AccessToken;
-            }
-            catch (BoxException ex)
+
+            while (true)
             {
-                var serverDate = ex.ResponseHeaders.Date;
-                if (serverDate.HasValue)
+                try
                 {
-                    var date = serverDate.Value;
-                    assertion = ConstructJWTAssertion(subId, subType, date.LocalDateTime);
                     result = JWTAuthPost(assertion);
                     return result.AccessToken;
                 }
-                else
+                catch (BoxException ex)
                 {
-                    throw ex;
-                }
+                    //need to wait for Retry-After seconds and then retry request
+                    var retryAfterHeader = ex.ResponseHeaders != null ? ex.ResponseHeaders.RetryAfter : null;
+
+                    // If we get a retryable/transient error code and this is not a multi part request (meaning a file upload, which cannot be retried
+                    // because the stream cannot be reset) and we haven't exceeded the number of allowed retries, then retry the request.
+                    // If we get a 202 code and has a retry-after header, we will retry after
+                    if ((ex.StatusCode == HttpRequestHandler.TooManyRequests
+                        ||
+                        ex.StatusCode == HttpStatusCode.InternalServerError
+                        ||
+                        ex.StatusCode == HttpStatusCode.BadGateway
+                        ||
+                        ex.StatusCode == HttpStatusCode.ServiceUnavailable
+                        ||
+                        ex.StatusCode == HttpStatusCode.GatewayTimeout
+                        ||
```

Comment:
May want to retry on the 400 error where the Unix time on the user's machine does not match the server.

Reply:
Done!